<!DOCTYPE html
	PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
	 "http://www.w3.org/TR/html4/loose.dtd">
<html><head><title>KEGG GENOME: Chlamydia trachomatis L1/1322/p2</title>
<meta name="content" content="text/html; charset=utf-8">
<meta name="http-equiv" content="Content-Type">
<link rel="stylesheet" type="text/css" href="/css/kegg2.css">
<link rel="stylesheet" type="text/css" href="/kegg/js/menubar/menubar.css">
<style type="text/css">
<!-- 
li,ul {margin-left: 10px;padding: 0px;list-style: none;}
 .object { display: inline; }
 -->
</style>
<script src="/kegg/js/menubar/menubar.js" type="text/JavaScript"></script>
</head>
<body bgcolor="#ffffff" link="#003399" vlink="#003399">
<table width="820">
<tr>
<td><a href="/kegg/kegg2.html"><img align="middle" alt="KEGG" border="0" src="/Fig/bget/kegg1.gif"></a>&nbsp;&nbsp;&nbsp;<font class="title1">Chlamydia trachomatis L1/1322/p2</font></td>
</tr>
</table>

<br>
<table width="820" cellpadding="2" class=frame1>
<tr><td class=td10 nowrap>
 <table width="820">
<tr><td>
<th><font class="current">Genome info</font></th><th><a href="./show_organism?menu_type=pathway_maps&amp;org=ctrp">Pathway map</a></th><th><a href="./show_organism?menu_type=gene_catalogs&amp;org=ctrp">Brite hierarchy</a></th><th><a href="./show_organism?menu_type=pathway_modules&amp;org=ctrp">Module</a></th>  <th><a href="/kegg-bin/show_genomemap_top?org_id=ctrp">Genome map</a></th>
  <th><a href="http://www.genome.jp/tools-bin/search_sequence?prog=blast&db=ctrp&seqid=aa">Blast</a></th>
  <th><a href="/kegg-bin/get_htext?query=ctrp&amp;htext=br08601.keg&amp;option=-a">Taxonomy</a></th>

</td></tr>
 </table>
</td></tr>

<tr><td>
<!-- start -->
<table width="800" cellpadding="10" align="center">
<tr><td>
<div style="float:left">
<form name="bfind" method="get" action="/dbget-bin/www_bfind_sub">
  Search genes:
  <input type="hidden" name="max_hit" value="1000">
  <input type="hidden" name="dbkey" value="ctrp">
  <input type="hidden" name="mode" value="bfind">
  <input type="text" size=40 name="keywords">
  <input type="submit" value="Go">
  <input type="reset" value="Clear">
</form>
</div>
<div style="clear:both"></div>
<hr class=frame1>
<h4 class="title10">Genome information</h4><table cellpadding=5 cellspacing=0><tr><td nowrap valign="top"><b>T number</b></td><td><a href='/dbget-bin/www_bget?gn:T02557'>T02557</a></td></tr>
<tr><td nowrap valign="top"><b>Org code</b></td><td>ctrp</td></tr>
<tr><td nowrap valign="top"><b>Aliases</b></td><td>1075087</td></tr>
<tr><td nowrap valign="top"><b>Full name</b></td><td>Chlamydia trachomatis L1/1322/p2</td></tr>
<tr><td nowrap valign="top"><b>Definition</b></td><td>Chlamydia trachomatis L1/1322/p2</td></tr>
<tr><td nowrap valign="top"><b>Annotation</b></td><td>KOALA</td></tr>
<tr><td nowrap valign="top"><b>Taxonomy</b></td><td>TAX: <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=1075087">1075087</a></td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Lineage</b></td><td>Bacteria; Chlamydiae; Chlamydiales; Chlamydiaceae; Chlamydia/Chlamydophila group; Chlamydia</td></tr>
<tr><td nowrap valign="top"><b>Data source</b></td><td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/319/165/GCA_000319165.1_ASM31916v1">GenBank</a> (Assembly: <a href="https://www.ncbi.nlm.nih.gov/assembly/GCA_000319165.1">GCA_000319165.1</a>)<br/>            BioProject: <a href='https://www.ncbi.nlm.nih.gov/bioproject/71461'>71461</a></td></tr>
<tr><td nowrap valign="top"><b>Original DB</b></td><td><a href="http://www.sanger.ac.uk/resources/downloads/bacteria/chlamydia-trachomatis.html">Sanger</a></td></tr>
<tr><td nowrap valign="top"><b>Chromosome</b></td><td>Circular</td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Sequence</b></td><td>GB: <a href="https://www.ncbi.nlm.nih.gov/nuccore/HE601951">HE601951</a></td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Length</b></td><td>1039428</td></tr>
<tr><td nowrap valign="top"><b>Plasmid</b></td><td>pL11322; Circular</td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Sequence</b></td><td>GB: <a href="https://www.ncbi.nlm.nih.gov/nuccore/HE603219">HE603219</a></td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Length</b></td><td>7500</td></tr>
<tr><td nowrap valign="top"><b>Statistics</b></td><td>Number of nucleotides:       1046928<br>            Number of protein genes:         898<br>            Number of RNA genes:              44<br></td></tr>
<tr><td nowrap valign="top"><b>Created</b></td><td>2013</td></tr>
<tr><td nowrap valign="top"><b>Reference</b></td><td>PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22406642">22406642</a></td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Authors</b></td><td>Harris SR et al.</td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Title</b></td><td>Whole-genome analysis of diverse Chlamydia trachomatis strains identifies phylogenetic relationships masked by current clinical typing.</td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Journal</b></td><td>Nat Genet 44:413-9, S1 (2012)<br/>            DOI: <a href='https://doi.org/10.1038/ng.2214'>10.1038/ng.2214</a></td></tr>
</table>
</td></tr>
</table>
<br>
<!-- end -->
</td></tr>

<tr><td align="center" class=td10>
  <table width=730>
  <tr>
	<th><a href="/kegg/kegg2.html">KEGG2</a></th>
	<th><a href="/kegg/pathway.html">PATHWAY</a></th>
	<th><a href="/kegg/brite.html">BRITE</a></th>
	<th><a href="/kegg/genes.html">GENES</a></th>
	<th><a href="/kegg/genome.html">GENOME</a></th>
	<th><a href="/kegg/ligand.html">LIGAND</a></th>
	<th><a href="/kegg/disease/">DISEASE</a></th>
	<th><a href="/kegg/drug/">DRUG</a></th>
  </tr>
  </table>
</td></tr>
</table>

</body>
</html>