<!DOCTYPE html
	PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
	 "http://www.w3.org/TR/html4/loose.dtd">
<html><head><title>KEGG GENOME: Leptospira interrogans serovar Linhai</title>
<meta name="content" content="text/html; charset=utf-8">
<meta name="http-equiv" content="Content-Type">
<link rel="stylesheet" type="text/css" href="/css/kegg2.css">
<link rel="stylesheet" type="text/css" href="/kegg/js/menubar/menubar.css">
<style type="text/css">
<!-- 
li,ul {margin-left: 10px;padding: 0px;list-style: none;}
 .object { display: inline; }
 -->
</style>
<script src="/kegg/js/menubar/menubar.js" type="text/JavaScript"></script>
</head>
<body bgcolor="#ffffff" link="#003399" vlink="#003399">
<table width="820">
<tr>
<td><a href="/kegg/kegg2.html"><img align="middle" alt="KEGG" border="0" src="/Fig/bget/kegg1.gif"></a>&nbsp;&nbsp;&nbsp;<font class="title1">Leptospira interrogans serovar Linhai</font></td>
</tr>
</table>

<br>
<table width="820" cellpadding="2" class=frame1>
<tr><td class=td10 nowrap>
 <table width="820">
<tr><td>
<th><font class="current">Genome info</font></th><th><a href="./show_organism?menu_type=pathway_maps&amp;org=lis">Pathway map</a></th><th><a href="./show_organism?menu_type=gene_catalogs&amp;org=lis">Brite hierarchy</a></th><th><a href="./show_organism?menu_type=pathway_modules&amp;org=lis">Module</a></th>  <th><a href="/kegg-bin/show_genomemap_top?org_id=lis">Genome map</a></th>
  <th><a href="http://www.genome.jp/tools-bin/search_sequence?prog=blast&db=lis&seqid=aa">Blast</a></th>
  <th><a href="/kegg-bin/get_htext?query=lis&amp;htext=br08601.keg&amp;option=-a">Taxonomy</a></th>

</td></tr>
 </table>
</td></tr>

<tr><td>
<!-- start -->
<table width="800" cellpadding="10" align="center">
<tr><td>
<div style="float:left">
<form name="bfind" method="get" action="/dbget-bin/www_bfind_sub">
  Search genes:
  <input type="hidden" name="max_hit" value="1000">
  <input type="hidden" name="dbkey" value="lis">
  <input type="hidden" name="mode" value="bfind">
  <input type="text" size=40 name="keywords">
  <input type="submit" value="Go">
  <input type="reset" value="Clear">
</form>
</div>
<div style="clear:both"></div>
<hr class=frame1>
<h4 class="title10">Genome information</h4><table cellpadding=5 cellspacing=0><tr><td nowrap valign="top"><b>T number</b></td><td><a href='/dbget-bin/www_bget?gn:T03859'>T03859</a></td></tr>
<tr><td nowrap valign="top"><b>Org code</b></td><td>lis</td></tr>
<tr><td nowrap valign="top"><b>Aliases</b></td><td>1395589</td></tr>
<tr><td nowrap valign="top"><b>Full name</b></td><td>Leptospira interrogans serovar Linhai</td></tr>
<tr><td nowrap valign="top"><b>Definition</b></td><td>Leptospira interrogans serovar Linhai 56609</td></tr>
<tr><td nowrap valign="top"><b>Annotation</b></td><td>KOALA</td></tr>
<tr><td nowrap valign="top"><b>Taxonomy</b></td><td>TAX: <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=1395589">1395589</a></td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Lineage</b></td><td>Bacteria; Spirochaetes; Leptospirales; Leptospiraceae; Leptospira</td></tr>
<tr><td nowrap valign="top"><b>Data source</b></td><td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/941/035/GCA_000941035.1_ASM94103v1">GenBank</a> (Assembly: <a href="https://www.ncbi.nlm.nih.gov/assembly/GCA_000941035.1">GCA_000941035.1</a>)<br/>            BioProject: <a href='https://www.ncbi.nlm.nih.gov/bioproject/217894'>217894</a></td></tr>
<tr><td nowrap valign="top"><b>Keywords</b></td><td>Human pathogen</td></tr>
<tr><td nowrap valign="top"><b>Disease</b></td><td><a href='/dbget-bin/www_bget?ds:H00355'>H00355</a>  Leptospirosis</td></tr>
<tr><td nowrap valign="top"><b>Comment</b></td><td>Initially isolated from a human patient with leptospirosis in Linhai, Zhejiang Province in 1954.<br></td></tr>
<tr><td nowrap valign="top"><b>Chromosome</b></td><td>1; Circular</td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Sequence</b></td><td>GB: <a href="https://www.ncbi.nlm.nih.gov/nuccore/CP006723">CP006723</a></td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Length</b></td><td>4331770</td></tr>
<tr><td nowrap valign="top"><b>Chromosome</b></td><td>2; Circular</td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Sequence</b></td><td>GB: <a href="https://www.ncbi.nlm.nih.gov/nuccore/CP006724">CP006724</a></td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Length</b></td><td>404857</td></tr>
<tr><td nowrap valign="top"><b>Plasmid</b></td><td>lcp1; Circular</td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Sequence</b></td><td>GB: <a href="https://www.ncbi.nlm.nih.gov/nuccore/CP006725">CP006725</a></td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Length</b></td><td>67282</td></tr>
<tr><td nowrap valign="top"><b>Plasmid</b></td><td>lcp2; Circular</td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Sequence</b></td><td>GB: <a href="https://www.ncbi.nlm.nih.gov/nuccore/CP006726">CP006726</a></td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Length</b></td><td>56757</td></tr>
<tr><td nowrap valign="top"><b>Plasmid</b></td><td>lcp3; Circular</td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Sequence</b></td><td>GB: <a href="https://www.ncbi.nlm.nih.gov/nuccore/CP006727">CP006727</a></td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Length</b></td><td>54986</td></tr>
<tr><td nowrap valign="top"><b>Statistics</b></td><td>Number of nucleotides:       4915652<br>            Number of protein genes:        4141<br>            Number of RNA genes:              42<br></td></tr>
<tr><td nowrap valign="top"><b>Created</b></td><td>2015</td></tr>
<tr><td nowrap valign="top"><b>Reference</b></td><td>PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25887950">25887950</a></td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Authors</b></td><td>Zhu W et al.</td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Title</b></td><td>Identification of three extra-chromosomal replicons in Leptospira pathogenic strain and development of new shuttle vectors.</td></tr>
<tr><td nowrap valign="top"><b>&nbsp;&nbsp;&nbsp;&nbsp;Journal</b></td><td>BMC Genomics 16:90 (2015)<br/>            DOI: <a href='https://doi.org/10.1186/s12864-015-1321-y'>10.1186/s12864-015-1321-y</a></td></tr>
</table>
</td></tr>
</table>
<br>
<!-- end -->
</td></tr>

<tr><td align="center" class=td10>
  <table width=730>
  <tr>
	<th><a href="/kegg/kegg2.html">KEGG2</a></th>
	<th><a href="/kegg/pathway.html">PATHWAY</a></th>
	<th><a href="/kegg/brite.html">BRITE</a></th>
	<th><a href="/kegg/genes.html">GENES</a></th>
	<th><a href="/kegg/genome.html">GENOME</a></th>
	<th><a href="/kegg/ligand.html">LIGAND</a></th>
	<th><a href="/kegg/disease/">DISEASE</a></th>
	<th><a href="/kegg/drug/">DRUG</a></th>
  </tr>
  </table>
</td></tr>
</table>

</body>
</html>